ARMmbed/mbed-os
# Nordic BLE: Allow configuration of softdevice parameters (old)

## Description

Allow configuration (via defines) of some of the key settings for the nordic softdevice.
* CENTRAL_LINK_COUNT
* PERIPHERAL_LINK_COUNT
* gatts_enable_params.attr_tab_size
* gatts_enable_params.service_changed
* common_enable_params.vs_uuid_count

These settings control the range of functionality enabled in the softdevice as well as ram consumption.
In particular, reducing these values is critical to enable usage of 16K nrf51 devices.

The defaults have not been changed so this should not effect existing users.

The values can be adjusted on a per-project basis by adding some build defines or `mbed_app.json` configs, eg:
``` json
{
    "config": {
        "ble_central_link_count": {
            "help": "When acting as BLE central, how many peripherals can connect",
            "macro_name": "CENTRAL_LINK_COUNT",
            "value": "0"
        },
        "ble_peripheral_link_count": {
            "help": "When acting as BLE peripheral, how many centrals can we conenct to",
            "macro_name": "PERIPHERAL_LINK_COUNT",
            "value": "1"
        },
        "ble_gatt_attr_tab_size": {
            "help": "The size of the table used to hold gatts. Can be adjusted by trial and error",
            "macro_name": "GATTS_ATTR_TAB_SIZE",
            "value": "0x390"
        },
        "ble_uuid_table_max_entries": {
            "help": "maximum number of 128-bit UUIDs with distinct bases that we expect to be in use",
            "macro_name": "UUID_TABLE_MAX_ENTRIES",
            "value": "2"
        }
    }
}
```

## Status

**READY**

## Migrations

There should be no changes to end user code.

NO

## Related PRs

None

## Steps to test or reproduce

When debugging code, a breakpoint can be set in `softdevice_enable()` function in `mbed-os/targets/TARGET_NORDIC/TARGET_NRF5/TARGET_SDK11/softdevice/common/softdevice_handler/softdevice_handler.c` 

Before `sd_ble_enable()` has been run in that function, `ram_start` variable holds the current length of ram being allocated for the softdevice. 
After `sd_ble_enable()`, `app_ram_base` variable is updated to hold the amount required by the softdevice for the current configuration. 

Changing the values above in defines/config should result in a change of ram requirements.

For use on smaller parts, once the values have been reduced and the softdevices' requirements have been measured as above, the RAM: ORIGIN and LENGTH definitions can be updated to suit in the linker file; `NRF51822.ld` (or equivalent)
